Add CEP for Repodata Wheel Support

danyeaw opened the pull request:
Updated version to replace #144, developed with @travishathaway.

This CEP outlines how native support for pure Python wheel packages could be achieved by adding support for them in repodata. When implemented, conda clients will be able to seamlessly install conda packages and pure Python wheels from enabled channels.

## Checklist for submitter

- [X] I am submitting a new CEP: [Repodata Wheel Support](https://github.com/danyeaw/ceps/blob/wheel-repodata/cep-XXXX.md).
  - [X] I am using the CEP template by creating a copy `cep-0000.md` named `cep-XXXX.md` in the root level.
- [ ] I am submitting modifications to CEP XX. <!-- reflect CEP number here -->
- [ ] Something else: (add your description here).

<!-- delete section below if this is not a new CEP -->
## Checklist for CEP approvals

- [ ] The vote period has ended and the vote has passed the necessary quorum and approval thresholds.
- [ ] A new CEP number has been minted. Usually, this is `${greatest-number-in-main} + 1`.
- [ ] The `cep-XXXX.md` file has been renamed accordingly.
- [ ] The `# CEP XXXX - ` header has been edited accordingly.
- [ ] The CEP status in the table has been changed to approved.
- [ ] The last modification date in the table has been updated accordingly.
- [ ] The table in the README has been updated with the new CEP entry.
- [ ] The `pre-commit` checks are passing.


<!-- Just as a reminder, everyone in all conda org spaces (including PRs)
     must follow the Conda Org Code of Conduct (link below).

     Finally, once again, thanks for your time and effort. If you have any
     feedback in regards to your experience contributing here, please
     let us know!

     Helpful links:
       - Conda Org COC: https://github.com/conda/governance/blob/main/CODE_OF_CONDUCT.md
-->

danyeaw committed:
Add CEP for Repodata Wheel Support

Co-authored-by: Travis Hathaway <travis.j.hathaway@gmail.com>

jjhelmus commented on cep-XXXX.md:
```diff
@@ -0,0 +1,561 @@
+
+### conda-pupa creates on-the-fly conversion plugin (July 2024)
+
+Daniel Holth creates a conda plugin that supports on-the-fly conversion of conda packages to wheels in a local channel.
```
This seems to be leaving out some history

pavelzw commented on cep-XXXX.md:
```diff
@@ -0,0 +1,561 @@
+
+### Pixi Integrates with uv (Jan 2024)
+
+Pixi changes course to use uv directly instead of rip, which unlocks features like editable installations, and git and path dependencies.
```
these are all now available for conda-only workflows through pixi-build.

danyeaw replied:
Hey @pavelzw, thanks so much for the feedback! Do you think we are missing a milestone in our brief history section? This pixi build feature is more about building path/git for conda packages than installing wheels isn't it?

danyeaw committed:
Add conda-whl-channel to the history timeline

danyeaw committed:
Grammar fixup

danyeaw committed:
Apply consistency and style updates

Co-authored-by: Travis Hathaway <travis.j.hathaway@gmail.com>

danyeaw committed:
Fix lint errors

beckermr commented:
I'm going to leave my opinions on the general goals/ideas/features of this CEP in an effort to help bring other perspectives to this CEP. 

TL;DR - As a conda-forge/core developer, I personally would NOT recommend folks use this feature, nor would I enable or offer support for this feature for conda-forge.

The CEP states

> By adding native support for pure Python wheels to repodata, conda clients can:
> 
> - Resolve dependencies across conda and PyPI packages in a single solve
> - Provide users with transparent access to the broader Python ecosystem
> - Maintain environment consistency and reproducibility
> - Eliminate the cognitive burden of managing two package managers
> - Fill gaps in conda package availability without requiring new conda builds
> - Reduce the maintenance burden by fully or partially eliminating the need to create and maintain conda recipes for pure-Python packages
> 

Here is a point-by-point explanation of why in my estimation this feature simply would not work for conda-forge. 

> Resolve dependencies across conda and PyPI packages in a single solve

In a non-trivial fraction of cases, even for pure-Python packages, the requirements in the conda-forge have subtle differences from the upstream requirements. These changes range from package renaming (e.g., `-` and `_` are identical for python packages, but not conda packages, matplotlib vs matplotlib-base, etc.) to more substantial changes that enable broader compatibility (e.g., upstream uses exact pins for dependencies, but conda-forge relaxes them because they are clearly not needed). I can imagine that when it comes to pure-Python packages that depend on a compiled backend package, things will also be funky in some cases. 

These requirement differences will likely lead to some funky solves that either the conda or conda-forge developers will hear about.

> Maintain environment consistency and reproducibility

Environment consistency is a tricky concept IMHO. For sure with this CEP one can in some cases create an environment where all constraints are satisfied. However, if the repodata is wrong, due to the issues outlined above, the formal consistency of the requirements doesn't really matter.

Reproducibility is an even trickier concept. For an environment to be reproducible, one needs to have the same solver with that solver run under the same conditions. Let's assume the conda version is fixed and the solver command is run on the same machine. Even then the constraint on having the same conditions combined with this CEP in effect means that both the upstream wheel metadata and the conda channel metadata have to be the same. Given that the most likely source of wheels is pypi, there is no way one can promise those same conditions. 

Even if we restrict ourselves to environments built from lock files, the combination of the conda channel with pypi as a source of wheels will also not always be reproducible. PyPI users can delete packages (as opposed to simply yanking packages) and those deletions will break even locked envs. We do not allow package deletions on conda-forge for this exact reason. Thus for conda-forge, we could not recommend using this feature for reproducible envs from lock files unless PyPI turns off the ability for users to delete files.

> Eliminate the cognitive burden of managing two package managers

The vast majority of the cognative burden is the differing and interacting repodata, not whether or not one types `pip install` or `conda install`. This CEP doesn't and cannot address that issue. Even consistently interpreting the repodata in a single solver doesn't address this issue.

> Fill gaps in conda package availability without requiring new conda builds
> Reduce the maintenance burden by fully or partially eliminating the need to create and maintain conda recipes for pure-Python packages

For the reasons stated above, I am personally skeptical injecting pure-Python wheel metadata into the repodata would consistently result in a correct-enough environments to eliminate the need for new conda builds or the need to repackage pure-python Packages in conda-forge. I am not saying this doesn't work some of the time. Instead I am saying that the solution proposed in this CEP is not so much better than the current "conda, then pip" solve that it can achieve the arguably difficult goals above. Stated another way, in a world where this feature existed instead of the feature to pip-install overtop of conda environments, conda-forge likely would still need/want to repackage everything.

### Other comments

> When there are naming differences between channels, wheel records MUST use the conda-forge package name as the standard.

This statement is a red-flag for me on this CEP. First, conda-forge itself doesn't have an authoritative mapping of its own packages back to python wheels. There are several approaches in the wild and none of them is standardized into an automated bit of repodata tools like conda/mamba/pixi/rattler can read and interact with. See the discussions of PURLS. Second, treating conda-forge as a special channel specifically in a CEP (as opposed to simply using it as a motivating usecase), is definitely IMHO the antithesis of what a CEP is supposed to be. conda is a set of tools and standards and should not be singling out any one purveyor of conda packages.

> !=X.Y.Z → Omit dependency (conda does not support version exclusions; see Limitations section)

I don't follow this comment. Can you clarify? For sure I have used this operator in the run section before (see, e.g., https://github.com/conda-forge/ngmix-feedstock/blob/main/recipe/meta.yaml#L30), At minimum any requirement that is `foo!=xyz` should be added into the constrains section of the repodata. If it is a the wheel's run section, one should add the package without any constraints in the run section of the repodata as well. Th combination of these two additions should achieve the same effect as "install foo, but not version xyz."

h-vetinari commented:
Not commenting on the whole CEP (I share much of @beckermr's reservations, but currently don't have a strong opinion), just one aspect that's important to get right IMO:

> I think having the wheel index own the mapping is still the right approach, what if we have the index declare which channel it is mapping names to. For example, we could add an optional field called `name_mapping_channel` to the info section like:

I think it would be a good idea to consider whether this can build on top of the proposed PEP [804](https://peps.python.org/pep-0804/) ([discourse](https://discuss.python.org/t/pep-804-an-external-dependency-registry-and-name-mapping-mechanism/103891)), which tries to standardize something useable around the whole name mapping issue.

It'd be certainly better if we can build on top of that (and help support that PEP) rather than inventing yet another scheme.

CC @jaimergp @rgommers @mgorny

danyeaw committed:
Updates to mapping, version exclusions, and PyPI deletion

Updates include:
- Clarifications on naming standards
- Channel mapping
- Patching capabilities for dependency management

# Conflicts:
#	cep-XXXX.md

danyeaw committed:
Add section for centralized name mapping in PEP804

danyeaw committed:
Small formatting fix

danyeaw commented:
Hi @beckermr, I made the updates to the relevant sections to fix version exclusions, remove reliance on conda-forge for channel mapping, clarify that we need downstream patching ability, added an implementation options section, and added recommendations for protecting against PyPI packages being deleted. Thanks again for all of the great feedback.

Hi @h-vetinari
> I think it would be a good idea to consider whether this can build on top of the proposed PEP [804](https://peps.python.org/pep-0804/)

Thanks that's a great point that we should build on this! I added a call out to PEP-804 in the [Naming standard and channel mapping](https://github.com/danyeaw/ceps/blob/wheel-repodata/cep-XXXX.md#naming-standard-and-channel-mapping) section.

danyeaw committed:
Fix code block not closed and missing URL

JeanChristopheMorinPerso commented on cep-XXXX.md:
```diff
@@ -0,0 +1,633 @@
+2. **Environment markers**:** Only Python version markers are converted to dependencies. Other environment markers (OS, platform, etc.) are ignored based on the pure Python assumption.
+3. **Conditionals and Extras:** Conditional dependencies and extras are specified by a separate CEP.
+4. **Repodata size:** Supporting a significant portion of pure Python packages from PyPI (potentially hundreds of thousands of packages with multiple versions each) will substantially increase repodata size. Channels adopting wheel support at scale will need to implement sharded repodata ([CEP 16][cep-16]) to maintain acceptable performance. Alternatively, channels may choose to curate a subset o ...
+5. **PyPI package deletion:** PyPI allows package maintainers to delete packages (as opposed to just yanking them), which can break locked environments that reference those packages. Channels using external PyPI URLs directly are subject to this risk. For production use and reproducible environments, channels SHOULD mirror and store wheel artifacts locally rather than relying solely on external P ...
```
I agree that deletion is sometimes a problem, but I'm wondering if this could encourage unsafe practices. Maybe we should add some wording or warnings regarding security. Or we could also remove the recommendation altogether or replace
>For production use and reproducible environments, channels SHOULD mirror and store wheel artifacts locally rather than relying solely on external PyPI URLs. This ensures that packages referenced in lock files remain available even if deleted from PyPI.

with
>For production use and reproducible environments, channels can mirror and store wheel artifacts locally rather than relying solely on external PyPI URLs.

(replace should with can, and remove the last sentence).

Opinions?

danyeaw replied:
Thanks for the feedback, I like those wording clarifications. I went with "MAY mirror" to match the keywords.

JeanChristopheMorinPerso commented on cep-XXXX.md:
```diff
@@ -0,0 +1,633 @@
+- Document any naming mappings specific to their channel
+- Ensure consistency with the declared channel's naming conventions
+
+Channel operators MUST document any naming conventions and mappings specific to their channel, regardless of whether `name_mapping_channel` is declared.
```
Are we confortable with saying `MUST` here, considering the current state of affairs when it comes to mappings in conda-forge and main? (and to be fair with conda-forge, they are in a better state than main).

danyeaw replied:
Thanks, I updated it from MUST to SHOULD.

JeanChristopheMorinPerso commented on cep-XXXX.md:
```diff
@@ -0,0 +1,633 @@
+
+- If `base_url` is defined in the repodata info object (per [CEP 15][cep-15]), `artifact_url` contains the path relative to `base_url`
+- If `base_url` is not defined, `artifact_url` contains the complete download URL
+- If `artifact_url` is not present (`null`), the download location is constructed from `base_url` and `fn` (existing behavior)
```
Minor: This is a bullet point under "When present, the `artifact_url` field follows these semantics". You could likely do this:
>When present, the `artifact_url` field follows these semantics:
>- If `base_url` is defined in the repodata info object (per [CEP 15][cep-15]), `artifact_url` contains the path relative to `base_url`
>- If `base_url` is not defined, `artifact_url` contains the complete download URL
>
>When not present, the download location is constructed from `base_url` and `fn` (existing behavior)

?

danyeaw replied:
Thanks so much for this feedback, I updated this based on your suggestion.

JeanChristopheMorinPerso commented on cep-XXXX.md:
```diff
@@ -0,0 +1,633 @@
+
+- Implement validation to ensure only pure Python wheels are included
+- Maintain a mapping of PyPI to conda-style names for their channel
+- Consider automation to keep wheel metadata synchronized with PyPI
```
I don't know if this is part of a standard or not, but files on PyPI are immutable. So it's impossible to modify the metadata of an existing wheel. Is it possible that you meant "automation to keep the repodata up-to-date with newer releases on PyPI"?

JeanChristopheMorinPerso reviewed (commented):
I left mostly questions

danyeaw committed:
Clarifications based on @JeanChristopheMorinPerso feedback

danyeaw committed:
Change from MUST to SHOULD

danyeaw committed:
Add clarification about using wheel metadata names

danyeaw commented on cep-XXXX.md:
```diff
@@ -0,0 +1,633 @@
+This CEP has the following known limitations:
+
+1. **Pure Python only:** This CEP explicitly does not address wheels with binary extensions, which require platform-specific compatibility guarantees beyond the current scope. Conda’s strength is binary compatibility, so using conda packages may be the optimal solution.
+2. **Environment markers**:** Only Python version markers are converted to dependencies. Other environment markers (OS, platform, etc.) are ignored based on the pure Python assumption.
```
Hi @JeanChristopheMorinPerso, since conditional dependencies is a separate CEP, what if we said that channel operators MAY patch and that they SHOULD support further conditional dependencies as they become available?

baszalmstra replied:
I think its important that we do describe how to convert those markers when conditional dependencies do exist. Because not all markers are trivial to convert (e.g. `platform_machine`, `python_full_version` vs `python_version`). I think that should be in this CEP but I can also imagine that a following could work.

danyeaw replied:
I added a non-normative reference to this for our conda-pypi conversion table.

baszalmstra commented on cep-XXXX.md:
```diff
@@ -0,0 +1,634 @@
+  - `<=X.Y.Z` → `<=X.Y.Z`
+  - `<X.Y.Z` → `<X.Y.Z`
+  - `>X.Y.Z` → `>X.Y.Z`
+  - `~=X.Y` → `>=X.Y,<X+1.0` (compatible release)
```
Conda also has this specifier.

danyeaw replied:
Thanks, TIL! I updated the proposal to include the `~=`.

baszalmstra commented on cep-XXXX.md:
````diff
@@ -0,0 +1,634 @@
+}
+```
+
+> **Note:** The `importlib-metadata` dependency is omitted because the `Requires-Python: >=3.8 makes its environment marker always false because the package requires Python >=3.8, making the python_version < '3.8' marker always false.
````
That's lucky in this case, maybe just remove this?

baszalmstra commented on cep-XXXX.md:
```diff
@@ -0,0 +1,634 @@
+
+- **Infrastructure burden:** Requires significant storage and bandwidth to host and serve converted packages that duplicate PyPI's CDN infrastructure, plus ongoing maintenance of the conversion pipeline and synchronization automation
+- **Resource inefficiency:** Wheels are already an excellent format for pure Python packages; conversion adds no technical value and wastes resources
+- **Timeliness:** Creates lag between PyPI publication and availability, with curation decisions needed about which packages to convert
```
This is not different from conversion of the wheels. A manual/automated step still needs to happen to convert the metadata. Converting the wheel data itself takes seconds at worst.

danyeaw replied:
As you point out, wheel conversion could also be automated without curation. I agree that this isn't a major differentiator between the approaches and I removed it. Thanks!

baszalmstra commented on cep-XXXX.md:
```diff
@@ -0,0 +1,634 @@
+- **Resource inefficiency:** Wheels are already an excellent format for pure Python packages; conversion adds no technical value and wastes resources
+- **Timeliness:** Creates lag between PyPI publication and availability, with curation decisions needed about which packages to convert
+- **Maintenance complexity:** Every wheel format change requires tooling updates; bugs affect all converted packages
+- **Ecosystem misalignment:** Converting away from wheels moves conda further from Python packaging standards
```
Worst case we keep the status quo in my eyes. Direct conversion from a pypi wheel to a conda package already removes the in between step of requiring a recipe.

wolfv commented on cep-XXXX.md:
```diff
@@ -0,0 +1,636 @@
+
+Examples:
+
+- `httpx-0.28.1`
```
I am a little bit afraid that this convention kinda closes the door on other wheel variants (like non-noarch packages) in the future.  We could just align it closer with the actual filename to future-proof things?

danyeaw committed:
Add clarifications from @baszalmstra and @wolfv

danyeaw commented on cep-XXXX.md:
```diff
@@ -0,0 +1,634 @@
+- `fn`: MUST be the wheel filename (e.g., package-1.0.0-py3-none-any.whl)
+- `subdir`: MUST be "noarch"
+- `noarch`: MUST be "python"
+- `artifact_url`: MUST be present and follow the semantics described above
```
Thanks, changed it to SHOULD.

danyeaw commented on cep-XXXX.md:
```diff
@@ -0,0 +1,634 @@
+Solvers MUST treat pure Python wheels as valid package candidates during dependency resolution with these constraints:
+
+- **Exclusivity:** Solvers MUST NOT install both a conda package and wheel for the same package name.
+- **Dependency satisfaction:** When a wheel is selected, its `depends` list MUST be satisfied like any conda package.
```
Thanks, I added this clarification!

danyeaw committed:
Reduce cons of wheel to conda package conversion

danyeaw commented on cep-XXXX.md:
```diff
@@ -0,0 +1,634 @@
+- Parse the new `packages.whl` section alongside existing package sections
+- Apply the same filtering and preference logic used for conda packages
+- Extract wheel metadata during solving to populate dependency information
+- Provide the ability to natively install wheels or on-the-fly convert wheels to conda packages for installation
```
This a great point, we have been more focused on getting the repodata to the right format and less on the native installation because our first phase is with conda-pypi which converts wheels to conda packages on the fly. A future step would be to support them natively. I'll start to work on a section about how this will work.

danyeaw replied:
Hey @baszalmstra and @jaimergp, I broke up the CEP in to 3:

1. Informational overview CEP about an overview, the history, and rejected ideas
2. CEP about repodata for wheels
3. CEP about client solving, mapping, and installation

Please let me know what you think.

danyeaw committed:
Rename mapping field to name_mappings

danyeaw committed:
Replace JSON name mapping with a parent channel

danyeaw committed:
Simplify version specifier language

jaimergp commented on cep-XXXX.md:
```diff
@@ -0,0 +1,613 @@
+
+## Motivation
+
+While conda remains a language-agnostic packaging distribution, installing packages for use with Python represents the majority of conda usage. Users frequently encounter packages only available as wheels on PyPI, forcing them to use hybrid workflows that mix conda and pip installations. This approach creates several problems:
```
```suggestion
While conda remains a language-agnostic packaging distribution, installing packages for use with Python represents the majority of conda usage. Users frequently encounter packages only available as wheels on PyPI, forcing them to use hybrid workflows that mix conda and PyPI installations. This approach creates several problems:
```

jaimergp commented on cep-XXXX.md:
```diff
@@ -0,0 +1,613 @@
+
+### Naming standard and channel mapping
+
+The key name for wheel records SHALL be constructed by combining the conda-style name (derived from METADATA and normalized per conda conventions) with the version, build string, and wheel tags (abi_tag-platform_tag) extracted from the filename. The python_tag is omitted as it is redundant with the build string. This ensures the name portion follows conda naming conventions, supports multiple reb ...
```
I don't see how we can extend the name-version-build syntax with extra fields without hurting backwards compatibility. All the wheel tags should be condensed in the build string, or dropped from the key.

danyeaw replied:
Hi @jaimergp, are you saying that we could limit client changes by making what was the "build" string a much longer set of wheel tags?

jaimergp commented on cep-XXXX.md:
```diff
@@ -0,0 +1,634 @@
+- Maintain a mapping of PyPI to conda-style names for their channel
+- Consider automation to keep the repodata up to date with newer releases on PyPI
+- Document any naming conventions specific to their channel
+- For production channels, consider mirroring wheel artifacts locally to ensure reproducibility and protect against PyPI deletions
```
I'm also thinking about this. We are adding a ton of metadata complexity to support the case where "maybe we don't want to copy this file over in a certain way". At that point, why not create a hybrid solver that knows how to traverse the PyPI index with certain rules (consider only pure Python wheels with dependencies already available, or only pure Python).

This proposal is a bit in the middle of nowhere: it requires preparsing all the wheel metadata and exposing it in a conda-friendly way, but then all the work is left for the client when it comes to installing the wheel. So we either go full conda and convert everything from the beginning, or stay 100% in PyPI world and integrate the solvers better (the mapping problem is present in all three solutions).

This proposal does one thing correctly though: it brings metadata control (by allowing repodata patching) without forcing anyone to copy any files (but in some cases it might be unavoidable).

jaimergp commented on cep-XXXX.md:
```diff
@@ -0,0 +1,634 @@
+
+### Add more conda packages
+
+Create and maintain new conda packages for each PyPI dependency needed. Tools like [Grayskull] exist to make this easier to convert. However, this is a significant workload for the community, with over half of all conda-forge packages being pure Python. Even with more dedicated resources, creating recipes for over 400 thousand pure Python packages is not achievable.
```
I could see a `conda-forge/wheels-index` repository where the automation pipelines are maintained folks add their requests to:

1. Add a project to the watchlist so it is included in the index
2. Yank certain wheels
3. Repodata patch
4. Import and archive pure wheel feedstocks
5. etc

I don't know if the resulting repodata would be added to conda-forge proper, or maybe a separate conda-forge-wheels channel (to keep the production channel lighter).

jaimergp reviewed (commented):
I have read the CEP and added a few comments. I am supportive of having better PyPI support in the conda ecosystem so users don't have to try their luck with multiple overlapping tools, but I have reservations with the scope and direction of this proposal.

First, it doesn't tell the whole story. The current CEP focuses on pre-processing wheel metadata in a friendly way for the conda solver. It doesn't go into the details and nuance of name mapping (with the complications it brings!), or what to do with the solved records once a solution is found: how to install the wheel, how to cache it, how to populate its `conda-meta/*.json` metadata. Ideally, the CEP would cover the whole story, or at least contextualize where it sits in the pipeline and which other CEPs to consult to get the full story.

I have reservations about the metadata-only approach too, and I think this needs to be better discussed in the Rejected ideas section. _Why_ is this approach so desirable compared to the others should go in a Rationale section. IIUC, this is because:

- Mirroring wheels is expensive and maybe unnecessary
- Metadata patching is desirable
- Easier to implement from the solver side of things

Also (unrelated, but worth exploring), if we do go this way, the proposal doesn't need to limit the strategy to wheels only. If we wanted to offer an agnostic field, like `packages.*`, where any package format could be allowed, it wouldn't take much more work:

- `url` shows where to download the artifact from
- `fn` informs of the format, and the client would know how to deal with it
- `depends` and others need to conform to conda conventions and name mappings

(This is to say that a big part of the specification right now doesn't seem to be very wheel-specific).

---

That aside, there are some editorial changes that would need to be made, but let's get there once we have agreed on a direction better.

danyeaw committed:
Updates from feedback from @jaimegrp

danyeaw commented on cep-XXXX.md:
```diff
@@ -0,0 +1,613 @@
+To help users understand which naming conventions are being used, channels MAY reference a parent channel that defines the naming conventions and mappings for wheel-to-conda name translation. Parent/child channel relationships will be defined in a separate CEP, which will specify how channels can declare a `parent_channel` field in the `info` section of repodata to inherit naming conventions and  ...
+For example, a wheel channel could reference conda-forge as its parent channel, inheriting conda-forge's established naming conventions and package name mappings. This allows channels to avoid duplicating name mapping definitions and ensures consistency across related channels.
+
+Channel operators SHOULD document any naming conventions and mappings specific to their channel, including whether they reference a parent channel for naming conventions.
```
Hi @jaimergp, we need to map the names of all the packages and the names of all the depends, otherwise the solver won't be able to create a solution. The PyPI name is saved as part of the fn field for the filename.

danyeaw commented on cep-XXXX.md:
```diff
@@ -0,0 +1,613 @@
+
+The package record uses conditional dependency syntax (`; if python < 3.9`) to declare `typing_extensions` only when the Python version is < 3.9, matching the original wheel METADATA. The Python version constraint of >=3.8 is directly mapped.
+
+## Rejected ideas
```
Nice catch, I forgot about that one. I added it as a rejected approach, thanks!

danyeaw committed:
Make dictionary structure more explicit

danyeaw commented on cep-XXXX.md:
```diff
@@ -0,0 +1,613 @@
+
+### Record structure and naming convention
+
+To support sparse repodata processing and maintain compatibility with conda's existing infrastructure, wheel records MUST use conda-style naming conventions rather than wheel filenames as keys. This allows solvers to efficiently filter packages by name without parsing the full package record.
```
Thanks, I updated this section.

danyeaw committed:
Move content to a rationale section

danyeaw commented on cep-XXXX.md:
```diff
@@ -0,0 +1,613 @@
+    - Example: `seaborn-base` (conda-forge) vs `seaborn` (PyPI)
+
+4. **Cross-channel naming:** Different conda channels may use different names
+    - Example: `pyperformance` (conda-forge) vs `performance` (main)
```
Thanks, I moved this to a rationale section.

danyeaw committed:
Update build string format

danyeaw committed:
Rename artifact_url to url

danyeaw commented on cep-XXXX.md:
```diff
@@ -0,0 +1,613 @@
+
+### Wheel download URLs
+
+This CEP introduces a new optional `artifact_url` field in package records to specify download locations for individual packages.
```
Thanks! I renamed this to `url`.

danyeaw committed:
Remove unnecessary fn field

danyeaw committed:
Reference separate URL field CEP draft

danyeaw committed:
Use backwards compatible repodata update strategy

danyeaw committed:
Reference the channel relations in repodata CEP draft

danyeaw committed:
Reference conditional and extras CEP

danyeaw committed:
Add extra_depends nomenclature

danyeaw committed:
Align fields: add fn, remove constrains, timestamp

danyeaw committed:
Reference download URL draft CEP

danyeaw commented on cep-XXXX.md:
```diff
@@ -0,0 +1,634 @@
+  - `!=X.Y.Z` → Add to `constrains` field
+- **Multiple specifiers:** Combine with commas (e.g., >=1.0,<2.0)
+- **Python version requirements:** Convert Requires-Python to explicit python dependency
+- **Environment markers:** Ignore markers other than Python version (pure Python assumption)
```
I updated the CEP to allow for additional environment markers. I also referenced a non-normative conversion table from conda-pypi.

danyeaw commented on cep-XXXX.md:
```diff
@@ -0,0 +1,636 @@
+- A shared relative or absolute `base_url` with all wheels in the same directory, by populating the `base_url` field and leaving the `artifact_url` field empty.
+- A manual PyPI repository with wheels in directories by the package name by populating the absolute URL in the `artifact_url` field, or the `base_url` and a relative path in the `artifact_url` field.
+- External PyPI mirrors or CDNs using absolute URLs by populating the `artifact_url` field, for example to <https://files.pythonhosted.org/packages/.../package-1.0.0-py3-none-any.whl>
+- Mixed sources within the same repodata file
```
This is now a separate CEP at https://github.com/conda/ceps/pull/151.

danyeaw committed:
Add details about the first pip subcommand

danyeaw committed:
Add informative overview CEP

danyeaw committed:
Cleanup wording

danyeaw committed:
More corrections for accuracy

danyeaw committed:
Update conditional and extras CEP PR links

danyeaw committed:
Add reference to conda-forge issue #26

intentionally-left-nil commented:
Hello, I'd like to introduce reroll [github](https://github.com/intentionally-left-nil/reroll), [pypi](https://pypi.org/project/py-reroll/) 
as a potential reference implementation to this CEP, if spec changes are made

Reroll's headline is that it is able to generate repodata.json for 99.9% of all supported wheels on pypi from my local analysis. There's more details in the README, but a supported wheel is an approximate of whether uv or pip would install the wheel in a python 3.4+ environment. For this initial implementation, I further scope it to CPython and manylinux, but those restrictions could be revisited later.

Not only does reroll convert the vast bulk of wheels, but it attempts to do so at full specificity, fully preserving the pypi semantics. Reroll will emit an error, rather than guessing or loosening a requirement it cannot translate.

Besides the actual code, I think this approach is important from a CEP perspective for several reasons.

* Ecosystem-wide analysis helps us understand if the proposals we are making will apply to the ecosystem, and how many edge cases or exceptions we would need to deal with
* Validating that a fully mechanical/deterministic approach can provide full fidelity, without the need to loosen or throw away restrictions to capture the semantic meaning from the wheel side
* Prove that repodata generation tools can fully handle converting all aspects of a wheel from PEP/pypi standards to conda, allowing this translation to be a requirement of the repodata itself. This allows solvers and other clients to just treat things as normal conda matchspec and related entities, greatly simplifying codepaths

# Changes needed
Reroll mostly conforms to the draft CEP, but differs in a few key areas.

1. Remove noarch-specific restrictions to the CEP, especially on the repodata side. All repodata should support any subdir, just like conda packages today. Instead, if the community is concerned about compiled wheel compatibility and other rabbit holes, I suggest adding text like: "Clients MAY choose to refuse to solve, or install a wheel which is not pure python. A client should detect this via the "fn" key to determine the original filename, for py*-none-any tags. A client which does choose to install a subdir-specific wheel does so without specific guidance from this CEP"
2. Add consistent wording that all fields of the v3/wheel directory are standard fields as according to the relevant conda CEP's. The wheel filename is in accordance to CEP-26. Build strings are also in accordance to CEP-26. Versions are according to CEP-29 and other related standards. All package names (the package and its dependencies) must be the correct conda-ecosystem name. Dependecies must be in accordance to matchspec CEP-29, and conda conventions (the python_abi metapackage, etc)
3. Remove, or vastly increase the specificity of the `### Dependency conversion` section. Reroll makes several decisions about dependency conversion which are aimed at its goals, and assumptions. For example, it does not follow the exact logic around operator conversion, but has its own logic. I would prefer to have this section be removed as a normative section of the CEP, and have the CEP instead focus on the outcome (valid matchspec), as opposed to the exact contents

## Why arch-specific repodata
Fundamentally, conda expresses its affinity to a particular platform via subdir. The modern `when` syntax, combined with virtual packages (such as `__win`) help approximate another path, but it is very easy to see where this falls down. For example, how does one translate the PEP-508 string of `packageA ; os_name != "nt"` There is no matchspec `[when="nt != ???"]` because matchspec here means this package has this version. Instead, conda expresses these things as subdirs, and wheels should as well.

Reroll shows how you can use partial solving to reduce PEP-508 markers into simpler conditions, which can then be expressed using the `when` syntax.

The other elephant in the room is for compiled wheels. One hill I will die on is that our users don't know or care about pure python wheels, `noarch` or other subdirs. They just want to conda install a package and have it work. Encoding "pure python" (which also doesn't even have a straightforward definition) means that our users will always have to understand "oh you can't get this wheel from conda for `reasons`". I think this is just very shortsighted. I understand all the complexities that come with this, but there are also plenty of wheels (rust, others) which link quite cleanly into environments and could be used. Instead, I think the decision of what types of wheels belong in the hands of channel administrators, and clients (to properly install the wheels).

I also don't think this is a decision that can be deferred until later. Properly handling "pure python" wheels with platform-specific dependencies requires at least discussing this topic. Although some dependencies can be solved with `when=[__win]` this is not comprehensive and quickly falls apart with complex requirements

## Why conda-specific fields
Clients don't need to understand PEP-440 version strings today. They don't need to understand PEP-508 markers, or pypi naming conventions. If we bleed any of this into the wheel-repodata it means that rattler, conda, and every other implementation will start needing to carry around (and agree upon) the proper parsing of these fields. Some examples: Version range parsing is slightly different between conda and pypi. This is why conda recipes usually have `python >=3.12,<3.13.0a0`. Also since matchspec doesn't support PEP-440 exactly, there would be cases where the version of your package can't be referenced by a dependency specifier. It's best just to treat everything as a normal conda package for the most part. In the same vein, conda currently places no real restrictions on the bulid tag. We should refrain from doing so here as well. Leave it up to packages and channels to decide the proper mechanism for their build tags. If people want to know the originating wheel for a repodata, the new `fn` tag already provides a mechanism to understand the source (and potentially take decisions based off it). Doing so this way will preserve flexibility in the future (for custom build tag matching) the same way that conda has afforded normal conda packages

# More information
If you're curious about the decisions behind reroll, please take a look at the [docs](https://github.com/intentionally-left-nil/reroll/tree/main/docs) section. I have spent as much time hand-writing and documenting my decisions as any other part of the repository and it contains a wealth of information about what and why. I also may have made mistakes, and it's much easier to point out mistakes in my documentation, so I can correct the code to match the spec.

If you want to know more about how I am analyzing reroll and conda-pypi across the pypi ecosystem, please look at [reroll-data](https://github.com/intentionally-left-nil/reroll-data). Please note the coding standards are much lower in that repository, and it's just my scratch space to explore and understand the ecosystem. I've been saving my (AI-assisted) analysis in the [notebooks](https://github.com/intentionally-left-nil/reroll-data/tree/main/notebooks) folder

jezdez commented:
For the record, @intentionally-left-nil's project `reroll` has not been reviewed by the conda maintainers team, and was developed independently. It should be considered a 3rd-party proposal at the moment for the CEP. I already communicated to Anil that I would have appreciated a little more heads-up before posting a claimed 99.9% support of supported wheels and that it seems to ignore the conda-pypi project.

danyeaw commented on cep-XXX1.md:
```diff
@@ -0,0 +1,377 @@
+
+- **`name`**: Taken from the wheel's METADATA `Name` field, normalized per [CEP 26][cep-26] and any name mappings aligned with the channel declared in `info.channel_relations.base` when present (see [Naming standard and channel mapping](#naming-standard-and-channel-mapping)).
+- **`version`**: Taken from the wheel's METADATA `Version` field, normalized per PEP 440.
+- **`build`**: Format `py{PY_MAJOR_VERSION}_{abi_tag}_{platform_tag}_{build_number}` (e.g., `py3_none_any_0`), conforming to CEP 26 build string conventions and the repodata record schema pattern `^([a-z0-9_.]+_)?[0-9]+$`. The build number MUST be at the end of the build string. The `{abi_tag}` and `{platform_tag}` are extracted from the wheel filename.
```
Hi @intentionally-left-nil, ok, I updated this to `{python_tag}`, thanks for the feedback!
